File: TRAQS Scheduling/TRAQS Scheduling/Services/Typography.swift
import SwiftUI

// MARK: - TRAQS Typography
// DM Sans across the entire app — five static weights bundled in /Fonts.
// All numerics use tabular figures (the `mono` helpers do this explicitly;
// the heading helpers inherit the same monospaced-digit treatment so totals
// and durations always column-align in cards and lists).

enum TFontName: String, CaseIterable {
    case regular   = "DMSans-Regular"
    case medium    = "DMSans-Medium"
    case semibold  = "DMSans-SemiBold"
    case bold      = "DMSans-Bold"
    case extrabold = "DMSans-ExtraBold"
}

extension TFontName {
    /// The face a web `fontWeight` resolves to.
    ///
    /// The web app styles by number and the apps ship five named files, so this is
    /// the translation — in ONE place, because two copies would drift and the
    /// symptom (type a half-weight off) is nearly invisible.
    ///
    /// Ranges rather than exact matches so an unlisted weight lands on its nearest
    /// shipped neighbour instead of falling through to a default that happens to be
    /// Regular.
    ///
    /// 300 (Light) is loaded by the web but not shipped here; it approximates to
    /// Regular. Add `DMSans-Light` the moment a ported screen actually uses it —
    /// see `WebFontWeightTests`.
    static func face(forWebWeight w: Int) -> TFontName {
        switch w {
        case ..<450: return .regular      // 300 (approximated), 400
        case ..<550: return .medium       // 500
        case ..<650: return .semibold     // 600
        case ..<750: return .bold         // 700
        default:     return .extrabold    // 800, and 900 — which the browser
                                          // clamps here too, since it never
                                          // loads a 900 face
        }
    }
}

enum TTypo {
    /// Brand wordmark — "traqs" 28pt ExtraBold, -0.04em letter-spacing.
    static func wordmark(_ size: CGFloat = 28) -> Font {
        Font.custom(TFontName.extrabold.rawValue, size: size, relativeTo: .largeTitle)
    }

    /// h1 (currently unused — header is replaced by the wordmark)
    static func h1(_ size: CGFloat = 32) -> Font {
        Font.custom(TFontName.bold.rawValue, size: size, relativeTo: .title)
    }

    /// h2 — hero numbers in cards
    static func h2(_ size: CGFloat = 26) -> Font {
        Font.custom(TFontName.bold.rawValue, size: size, relativeTo: .title2)
    }

    /// h3 — section titles, card titles
    static func h3(_ size: CGFloat = 22) -> Font {
        Font.custom(TFontName.bold.rawValue, size: size, relativeTo: .title3)
    }

    /// Body 15pt / 500
    static func body(_ size: CGFloat = 15) -> Font {
        Font.custom(TFontName.medium.rawValue, size: size, relativeTo: .body)
    }

    /// Body bold — for emphasized inline content
    static func bodyBold(_ size: CGFloat = 15) -> Font {
        Font.custom(TFontName.bold.rawValue, size: size, relativeTo: .body)
    }

    /// Secondary text — 14pt / 500
    static func sm(_ size: CGFloat = 14) -> Font {
        Font.custom(TFontName.medium.rawValue, size: size, relativeTo: .callout)
    }

    /// Secondary bold — 14pt / 700
    static func smBold(_ size: CGFloat = 14) -> Font {
        Font.custom(TFontName.bold.rawValue, size: size, relativeTo: .callout)
    }

    /// Caption / chip / label — 12pt / 600 (often UPPERCASE + 0.12em tracking)
    static func xs(_ size: CGFloat = 12) -> Font {
        Font.custom(TFontName.semibold.rawValue, size: size, relativeTo: .caption)
    }

    /// Caption bold — 12pt / 700
    static func xsBold(_ size: CGFloat = 12) -> Font {
        Font.custom(TFontName.bold.rawValue, size: size, relativeTo: .caption)
    }

    /// Tiny — 10pt / 700 (tab labels, badges)
    static func xxs(_ size: CGFloat = 10) -> Font {
        Font.custom(TFontName.bold.rawValue, size: size, relativeTo: .caption2)
    }

    /// "Mono" — DM Sans Medium with tabular figures, used wherever the desktop product
    /// used JetBrains Mono (timestamps, durations, percentages). We keep ONE typeface
    /// and rely on tabular-nums to keep digit columns aligned.
    static func mono(_ size: CGFloat = 12) -> Font {
        Font.custom(TFontName.medium.rawValue, size: size, relativeTo: .caption)
    }

    /// Mono bold — for tabular numbers that should stand out (hero totals, percentages).
    static func monoBold(_ size: CGFloat = 12) -> Font {
        Font.custom(TFontName.bold.rawValue, size: size, relativeTo: .caption)
    }
}

// MARK: - Convenience Text wrappers

extension View {
    /// Apply tabular-figure number rendering so digits column-align in lists/grids.
    /// Pair with `TTypo.mono*` for timestamps / durations / counts.
    func tnum() -> some View {
        self.monospacedDigit()
    }

    /// Uppercase + tracked label treatment (chips, section headers, "TRACKING", etc.).
    func tLabel(tracking: CGFloat = 1.4) -> some View {
        self.kerning(tracking).textCase(.uppercase)
    }
}

File: traqs-android/app/src/main/java/com/matrixsystems/traqs/ui/theme/Typography.kt
package com.matrixsystems.traqs.ui.theme

import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.sp

// Port of the iOS type scale (Services/Typography.swift).
//
// DM Sans throughout, five weights. Sizes and weights match iOS TTypo exactly so
// the same element is the same size on both platforms; iOS `tracking` is in
// points and Compose `letterSpacing` in sp is likewise absolute, so tracking
// values carry across 1:1 with no conversion.
//
// Where a caller needs a size other than the default, pass it — the defaults are
// the values iOS uses when it calls the helper bare.
object TTypo {
    // Brand wordmark — only the splash still sets type for this; the header uses
    // the image lockup.
    fun wordmark(size: TextUnit = 28.sp) = style(FontWeight.ExtraBold, size)

    // Display title at the top of a page. See PageTitle for the -4 tracking that
    // goes with it.
    fun h1(size: TextUnit = 32.sp) = style(FontWeight.Bold, size)

    // Hero numbers in cards.
    fun h2(size: TextUnit = 26.sp) = style(FontWeight.Bold, size)

    // Section titles, card titles.
    fun h3(size: TextUnit = 22.sp) = style(FontWeight.Bold, size)

    fun body(size: TextUnit = 15.sp) = style(FontWeight.Medium, size)
    fun bodyBold(size: TextUnit = 15.sp) = style(FontWeight.Bold, size)

    // Secondary text.
    fun sm(size: TextUnit = 14.sp) = style(FontWeight.Medium, size)
    fun smBold(size: TextUnit = 14.sp) = style(FontWeight.Bold, size)

    // Caption / chip / label. Usually uppercase with tracking — see `label`.
    fun xs(size: TextUnit = 12.sp) = style(FontWeight.SemiBold, size)
    fun xsBold(size: TextUnit = 12.sp) = style(FontWeight.Bold, size)

    // Tiny — tab labels, badges.
    fun xxs(size: TextUnit = 10.sp) = style(FontWeight.Bold, size)

    // "Mono" — DM Sans Medium standing in for the desktop product's JetBrains
    // Mono (timestamps, durations, percentages). ONE typeface app-wide; digit
    // columns stay aligned because DM Sans figures are tabular by default.
    fun mono(size: TextUnit = 12.sp) = style(FontWeight.Medium, size)
    fun monoBold(size: TextUnit = 12.sp) = style(FontWeight.Bold, size)

    // The uppercase tracked treatment iOS applies with `.tLabel(tracking:)`.
    // Compose has no textCase modifier, so CALL SITES must uppercase the string
    // themselves — this only carries the metrics.
    //
    // The tracking values in use, so a new label picks the right one rather than
    // inventing an eighth: 1.4 section headers · 1.0 in-card status labels ·
    // 0.8/0.6 button labels · 0.4 tag pills.
    fun label(size: TextUnit = 12.sp, tracking: TextUnit = 1.4.sp) =
        style(FontWeight.Bold, size).copy(letterSpacing = tracking)

    private fun style(weight: FontWeight, size: TextUnit) =
        TextStyle(fontFamily = DMSans, fontWeight = weight, fontSize = size)
}

// Tracking constants, named so the intent is readable at the call site.
object TTrack {
    val section = 1.4.sp    // SECTION HEADERS
    val status = 1.0.sp     // in-card status labels (TRACKING, PROGRESS)
    val button = 0.6.sp     // button labels
    val pill = 0.4.sp       // tag pills
}
